pkg/controller/replicaset/replica_set.go

```diff
@@ -650,3 +641,76 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
 	}
 	return manageReplicasErr
 }
+
+// slowStartBatch obtains total batch size to be processed and a function to execute during each batch.
+// It returns successfully processed batch size and an error if found.
+func slowStartBatch(totalBatchSize int, fn func() error) (int, error) {
+	var wg sync.WaitGroup
+	remainedBatchSize := totalBatchSize
+	totalProcessedBatchSize := 0
+	errCh := make(chan error, totalBatchSize)
+	var lock sync.Mutex
+	for batchSize := integer.IntMin(remainedBatchSize, controller.SlowStartInitialBatchSize); remainedBatchSize > 0; batchSize = integer.IntMin(2*batchSize, remainedBatchSize) {
+		wg.Add(batchSize)
+		for i := 0; i < batchSize; i++ {
+			go func() {
+				defer wg.Done()
+				err := fn()
+				if err != nil && !errors.IsTimeout(err) {
+					errCh <- err
+				} else if err == nil {
+					lock.Lock()
+					totalProcessedBatchSize++
+					lock.Unlock()
+				}
+
+			}()
+		}
+		wg.Wait()
+		remainedBatchSize -= batchSize
+	}
+
+	select {
+	case err := <-errCh:
+		// all errors have been reported before and they're likely to be the same, so we'll only return the first one we hit.
+		if err != nil {
+			return totalProcessedBatchSize, err
+		}
+	default:
+	}
+	return totalProcessedBatchSize, nil
+}
+
+func getCanAdoptFunc(rsc *ReplicaSetController, rs *extensions.ReplicaSet) func() error {
```

Comment:
No unit test is added for this function as (i think) it is covered by integration and e2e tests. This function only factors out the codes and does nothing else.